src/thermo/Phase.cpp

```diff
@@ -628,21 +717,40 @@ doublereal Phase::elementalMoleFraction(const size_t m) const
     return numerator / denom;
 }
 
-doublereal Phase::molarDensity() const
+double Phase::molarDensity() const
 {
     return density()/meanMolecularWeight();
 }
 
-void Phase::setMolarDensity(const doublereal molar_density)
+void Phase::setMolarDensity(const double molar_density)
 {
+    assertCompressible("setMolarDensity");
     m_dens = molar_density*meanMolecularWeight();
 }
 
-doublereal Phase::molarVolume() const
+double Phase::molarVolume() const
 {
     return 1.0/molarDensity();
 }
 
+void Phase::setDensity(const double density_)
+{
+    assertCompressible("setDensity");
+    if (density_ > 0.0) {
+        m_dens = density_;
+    } else {
+        throw CanteraError("Phase::setDensity()",
+            "density must be positive. density = {}", density_);
+    }
+}
+
+void Phase::setConstDensity(const double density_)
+{
+    if (density_ > 0.0) {
+        m_dens = density_;
+    }
+}
+
```

Comment:
I see the semantic purpose of distinguishing this function, but what is the practical purpose? Why remove the check of negative density? Presumably, someone could put a negative density in their input file...

Reply:
For incompressible substances, the `setDensity` path is blocked; `setConstDensity` is used instead to allow for assigning constant values.

Reply:
I renamed the function to `assignDensity`, and improved the docstring.